runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ParDoBoundMultiTranslator.java

```diff
@@ -112,8 +113,10 @@ public void translate(
             .collect(
                 Collectors.toMap(e -> e.getKey(), e -> ((PCollection<?>) e.getValue()).getCoder()));
 
-    boolean isStateful = DoFnSignatures.isStateful(transform.getFn());
-    final Coder<?> keyCoder = isStateful ? ((KvCoder<?, ?>) input.getCoder()).getKeyCoder() : null;
+    final Coder<?> keyCoder =
+        StateUtils.isStateful(transform.getFn())
```

Comment:
I guess we might as well as a getKeyCoder() in the stateUtils to avoid these logic.

Reply:
This logic to extract key coder from a class pipeline is only used here.